command/agent/config/config.go

```diff
@@ -169,14 +171,253 @@ func NewConfig() *Config {
 	}
 }
 
+// Merge merges two Agent configurations.
+func (c *Config) Merge(c2 *Config) *Config {
+	if c2 == nil {
+		return c
+	}
+
+	result := NewConfig()
+
+	result.SharedConfig = c.SharedConfig
+	if c2.SharedConfig != nil {
+		result.SharedConfig = c.SharedConfig.Merge(c2.SharedConfig)
+	}
+
+	result.AutoAuth = c.AutoAuth
+	if c2.AutoAuth != nil {
+		result.AutoAuth = c2.AutoAuth
+	}
+
+	result.Cache = c.Cache
+	if c2.Cache != nil {
+		result.Cache = c2.Cache
+	}
+
+	result.APIProxy = c.APIProxy
+	if c2.APIProxy != nil {
+		result.APIProxy = c2.APIProxy
+	}
+
+	result.DisableMlock = c.DisableMlock
+	if c2.DisableMlock {
+		result.DisableMlock = c2.DisableMlock
+	}
+
+	// For these, ignore the non-specific one and overwrite them all
+	result.DisableIdleConnsAutoAuth = c.DisableIdleConnsAutoAuth
+	if c2.DisableIdleConnsAutoAuth {
+		result.DisableIdleConnsAutoAuth = c2.DisableIdleConnsAutoAuth
+	}
+
+	result.DisableIdleConnsAPIProxy = c.DisableIdleConnsAPIProxy
+	if c2.DisableIdleConnsAPIProxy {
+		result.DisableIdleConnsAPIProxy = c2.DisableIdleConnsAPIProxy
+	}
+
+	result.DisableIdleConnsTemplating = c.DisableIdleConnsTemplating
+	if c2.DisableIdleConnsTemplating {
+		result.DisableIdleConnsTemplating = c2.DisableIdleConnsTemplating
+	}
+
+	result.DisableKeepAlivesAutoAuth = c.DisableKeepAlivesAutoAuth
+	if c2.DisableKeepAlivesAutoAuth {
+		result.DisableKeepAlivesAutoAuth = c2.DisableKeepAlivesAutoAuth
+	}
+
+	result.DisableKeepAlivesAPIProxy = c.DisableKeepAlivesAPIProxy
+	if c2.DisableKeepAlivesAPIProxy {
+		result.DisableKeepAlivesAPIProxy = c2.DisableKeepAlivesAPIProxy
+	}
+
+	result.DisableKeepAlivesTemplating = c.DisableKeepAlivesTemplating
+	if c2.DisableKeepAlivesTemplating {
+		result.DisableKeepAlivesTemplating = c2.DisableKeepAlivesTemplating
+	}
+
+	result.TemplateConfig = c.TemplateConfig
+	if c2.TemplateConfig != nil {
+		result.TemplateConfig = c2.TemplateConfig
+	}
+
+	for _, l := range c.Templates {
+		result.Templates = append(result.Templates, l)
+	}
+	for _, l := range c2.Templates {
+		result.Templates = append(result.Templates, l)
+	}
+
+	result.ExitAfterAuth = c.ExitAfterAuth
+	if c2.ExitAfterAuth {
+		result.ExitAfterAuth = c2.ExitAfterAuth
+	}
+
+	result.Vault = c.Vault
```

Comment:
The vault stanza could also be spread across multiple files.

Reply:
We don't support the Vault stanza (or any stanza) being spread across multiple files (which is how multi-config works for Vault itself). We will use the latest-defined Vault stanza.